pkg/datapath/linux/ipsec/probe_linux.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// +build linux
+
+package ipsec
+
+import (
+	"encoding/hex"
+	"errors"
+	"net"
+
+	"github.com/cilium/cilium/pkg/datapath/linux/linux_defaults"
+
+	"github.com/vishvananda/netlink"
+)
+
+const (
+	dummyIP  = "169.254.169.254"
+	aeadKey  = "4242424242424242424242424242424242424242"
+	aeadAlgo = "rfc4106(gcm(aes))"
+	stateId  = 42
+)
+
+func initDummyXfrmState() *netlink.XfrmState {
+	state := ipSecNewState()
+
+	k, _ := hex.DecodeString(aeadKey)
+	state.Aead = &netlink.XfrmStateAlgo{
+		Name:   aeadAlgo,
+		Key:    k,
+		ICVLen: 128,
+	}
+	state.Spi = int(stateId)
+	state.Reqid = stateId
+
+	state.Src = net.ParseIP(dummyIP)
+	state.Dst = net.ParseIP(dummyIP)
+	return state
+}
+
+func createDummyXfrmState(state *netlink.XfrmState) error {
+	state.Mark = &netlink.XfrmMark{
+		Value: linux_defaults.RouteMarkDecrypt,
+		Mask:  linux_defaults.IPsecMarkMaskIn,
+	}
+	state.OutputMark = &netlink.XfrmMark{
+		Value: linux_defaults.RouteMarkDecrypt,
+		Mask:  linux_defaults.RouteMarkMask,
+	}
+	return netlink.XfrmStateAdd(state)
+}
+
+// ProbeXfrmStateOutputMask probes the kernel to determine if it supports
+// setting the xfrm state output mask (Linux 4.19+). It returns an error if
+// the output mask is not supported or if an error occurred, nil otherwise.
+func ProbeXfrmStateOutputMask() error {
+	state := initDummyXfrmState()
+	err := createDummyXfrmState(state)
+	if err != nil {
+		return err
+	}
+	defer netlink.XfrmStateDel(state)
+
+	var probedState *netlink.XfrmState
+	if probedState, err = netlink.XfrmStateGet(state); err != nil {
+		return err
+	}
+	if probedState == nil || probedState.OutputMark == nil {
+		return errors.New("IPSec output mark attribute missing from xfrm probe")
+	}
+	if probedState.OutputMark.Mask != linux_defaults.RouteMarkMask {
+		return errors.New("Incorrect value for probed IPSec output mask attribute")
```

Comment:
Nit: errors strings should start with lowercase, except for abbreviations that should be capitalized (like above with IPSec).

Reply:
Oh? Sending the change in a few seconds, but why is the reason for that? They are generally not included as part of other sentences anyway.

Reply:
I go by https://github.com/golang/go/wiki/CodeReviewComments#error-strings. I think the main reason for this guideline is to make sure that error strings that are chained together throughout the code, as such for example:`unable to perform X: failed to delete Y: invalid input`.

While yes it's true in this case if we look at the callers of this function, we see that there is no such chain. However, I tend to follow the guidelines unless there's a very strong reason not to.

Reply:
Thanks! I'll read those guidelines tomorrow and will probably learn a bunch of things :student: